### PR TITLE
Draft: Update to Cosmos v0.40

### PR DESCRIPTION
**Pull request by migueldingli1997**
Opened for comparison purposes

### CODE REVIEW

File: cmd/ixod/cmd/root.go [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+	return rootCmd, encodingConfig
+}
+
+func initRootCmd(rootCmd *cobra.Command, encodingConfig params.EncodingConfig) {
+	authclient.Codec = encodingConfig.Marshaler
+
+	rootCmd.AddCommand(
+		genutilcli.InitCmd(app.ModuleBasics, app.DefaultNodeHome),
+		genutilcli.CollectGenTxsCmd(banktypes.GenesisBalancesIterator{}, app.DefaultNodeHome),
+		genutilcli.MigrateGenesisCmd(), //TODO gaia has custom function, make a custom function too?
```
Comment (migueldingli1997) [inline cmd/ixod/cmd/root.go:81]: Tackle or remove?

File: cmd/ixod/cmd/root.go [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+	skipUpgradeHeights := make(map[int64]bool)
+	for _, h := range cast.ToIntSlice(appOpts.Get(server.FlagUnsafeSkipUpgrades)) {
+		skipUpgradeHeights[int64(h)] = true
+	}
+
+	pruningOpts, err := server.GetPruningOptionsFromFlags(appOpts)
+	if err != nil {
+		panic(err)
+	}
+	//TODO not sure if these are necessary (but they are included in gaia)
```
Comment (migueldingli1997) [inline cmd/ixod/cmd/root.go:179]: Tackle or remove?

File: app/app.go [hunk 3]
```diff
@@ -479,7 +655,7 @@ func GetMaccPerms() map[string][]string {
 	return dupMaccPerms
 }
 
-func NewIxoAnteHandler(app *ixoApp) sdk.AnteHandler {
+func NewIxoAnteHandler(app *IxoApp, encodingConfig params.EncodingConfig) sdk.AnteHandler { //TODO encodingConfig
```
Comment (migueldingli1997) [inline app/app.go:658]: Tackle or remove TODO?

File: x/project/types/types.go [hunk 4]
```diff
@@ -3,14 +3,14 @@ package types
...
+	didexported "github.com/ixofoundation/ixo-blockchain/x/did/exported"
 )
 
 type (
-	InternalAccountID          string
-	AccountMap                 map[InternalAccountID]sdk.AccAddress
-	GenesisAccountMap          map[string]sdk.AccAddress
+	InternalAccountID string
+	//AccountMap                 map[InternalAccountID]sdk.AccAddress
+	//GenesisAccountMap          map[string]sdk.AccAddress
```
Comment (migueldingli1997) [inline x/project/types/types.go:13]: Remove these

File: x/project/types/params.go [hunk 5]
```diff
@@ -2,10 +2,14 @@ package types
...
 	sdk "github.com/cosmos/cosmos-sdk/types"
-	"github.com/cosmos/cosmos-sdk/x/params"
-	"github.com/ixofoundation/ixo-blockchain/x/did"
-	"github.com/ixofoundation/ixo-blockchain/x/ixo"
+	params "github.com/cosmos/cosmos-sdk/x/params/types"
+	didexported "github.com/ixofoundation/ixo-blockchain/x/did/exported"
+)
+
+const (
+	IxoNativeToken = "uixo"
```
Comment (migueldingli1997) [inline x/project/types/params.go:12]: Remove this and replace the usage of it with `ixotypes.IxoNativeToken`, with import `ixotypes "github.com/ixofoundation/ixo-blockchain/x/ixo/types"`

File: x/project/types/msgs.go [hunk 6]
```diff
@@ -116,21 +109,14 @@ func (msg MsgCreateProject) String() string {
 	return string(b)
 }
 
-func (msg MsgCreateProject) GetPubKey() string { return msg.PubKey }
+//func (msg MsgCreateProject) GetPubKey() string { return msg.PubKey }
```
Comment (migueldingli1997) [inline x/project/types/msgs.go:112]: Remove

File: x/project/genesis.go [hunk 7]
```diff
@@ -24,34 +27,34 @@ func InitGenesis(ctx sdk.Context, keeper Keeper, data GenesisState) {
...
 		accountMap := k.GetAccountMap(ctx, projectDoc.ProjectDid)
 		withdrawalInfo, _ := k.GetProjectWithdrawalTransactions(ctx, projectDoc.ProjectDid)
 
-		var subClaims []Claim
+		var subClaims types.Claims
 		claimIter := k.GetClaimIterator(ctx, projectDoc.ProjectDid)
 		for ; claimIter.Valid(); claimIter.Next() {
 			claim := k.MustGetClaimByKey(ctx, claimIter.Key())
-			subClaims = append(subClaims, claim)
+			subClaims = types.AppendClaims(subClaims, claim) //append(subClaims, claim)
```
Comment (migueldingli1997) [inline x/project/genesis.go:57]: Remove comment

File: x/project/auth.go [hunk 8]
```diff
@@ -2,70 +2,76 @@ package project
...
 
 // Identical to Cosmos DeductFees function, but tokens sent to project account
-func deductProjectFundingFees(bankKeeper bank.Keeper, ctx sdk.Context,
-	acc exported.Account, projectAddr sdk.AccAddress, fees sdk.Coins) error {
-	blockTime := ctx.BlockHeader().Time
-	coins := acc.GetCoins()
+func deductProjectFundingFees(bankKeeper bankkeeper.Keeper, ctx sdk.Context,
+	acc authtypes.AccountI, projectAddr sdk.AccAddress, fees sdk.Coins) error {
+	//blockTime := ctx.BlockHeader().Time
+	coins := bankKeeper.GetAllBalances(ctx, acc.GetAddress()) //acc.GetCoins()
```
Comment (migueldingli1997) [inline x/project/auth.go:74]: Comments to be removed

File: x/project/auth.go [hunk 9]
```diff
@@ -101,13 +107,16 @@ func getProjectCreationSignBytes(ctx sdk.Context, tx auth.StdTx, acc exported.Ac
...
 	)
 }
 
-func NewProjectCreationAnteHandler(ak auth.AccountKeeper, supplyKeeper supply.Keeper,
-	bk bank.Keeper, didKeeper did.Keeper, pubKeyGetter ixo.PubKeyGetter) sdk.AnteHandler {
+func NewProjectCreationAnteHandler(
+	ak authkeeper.AccountKeeper, bk bankkeeper.Keeper, didKeeper didkeeper.Keeper, signModeHandler signing.SignModeHandler,
+	pubKeyGetter ixotypes.PubKeyGetter) sdk.AnteHandler {
+	//ak auth.AccountKeeper, supplyKeeper supply.Keeper,
+	//bk bank.Keeper, didKeeper did.Keeper, pubKeyGetter ixo.PubKeyGetter) sdk.AnteHandler {
```
Comment (migueldingli1997) [inline x/project/auth.go:119]: Comments to be removed

File: x/payments/types/payments.go [hunk 10]
```diff
@@ -130,7 +112,8 @@ func (pc PaymentContract) Validate() error {
 	}
 
 	// Validate recipient distribution
-	if err := pc.Recipients.Validate(); err != nil {
+	var recipients Distribution = pc.Recipients
+	if err := recipients.Validate(); err != nil {
```
Comment (migueldingli1997) [inline x/payments/types/payments.go:116]:
```suggestion
	if err := Distribution(pc.Recipients).Validate(); err != nil {
```

File: x/payments/types/payments.go [hunk 11]
```diff
@@ -58,34 +52,22 @@ func (pt PaymentTemplate) Validate() error {
 	}
 
 	// Validate discounts
-	if err := pt.Discounts.Validate(); err != nil {
+	var discounts Discounts = pt.Discounts
+	if err := discounts.Validate(); err != nil {
```
Comment (migueldingli1997) [inline x/payments/types/payments.go:56]:
```suggestion
	if err := Discounts(pt.Discounts).Validate(); err != nil {
```

File: x/did/types/codec.go [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+func RegisterLegacyAminoCodec(cdc *codec.LegacyAmino) {
+	cdc.RegisterInterface((*exported.DidDoc)(nil), nil)
+
+	cdc.RegisterConcrete(&MsgAddDid{}, "did/AddDid", nil)
+	cdc.RegisterConcrete(&MsgAddCredential{}, "did/AddCredential", nil)
+
+	// TODO: https://github.com/ixofoundation/ixo-blockchain/issues/76
+	cdc.RegisterConcrete(&BaseDidDoc{}, "did/BaseDidDoc", nil)
+	//cdc.RegisterConcrete(DidCredential{}, "did/DidCredential", nil)
+	//cdc.RegisterConcrete(Claim{}, "did/Claim", nil)
```
Comment (migueldingli1997) [inline x/did/types/codec.go:21]: Remove comments

File: x/did/exported/did.go [hunk 13]
```diff
@@ -117,8 +121,8 @@ func (id IxoDid) String() string {
 }
 
 func VerifyKeyToAddr(verifyKey string) sdk.AccAddress {
-	var pubKey ed25519tm.PubKeyEd25519
-	copy(pubKey[:], base58.Decode(verifyKey))
+	var pubKey ed25519.PubKey //ed25519tm.PubKeyEd25519
```
Comment (migueldingli1997) [inline x/did/exported/did.go:124]: Remove comment?

File: x/payments/keeper/subscriptions.go [hunk 14]
```diff
@@ -3,53 +3,53 @@ package keeper
 import (
 	"fmt"
 	sdk "github.com/cosmos/cosmos-sdk/types"
-	"github.com/ixofoundation/ixo-blockchain/x/payments/internal/types"
+	types2 "github.com/ixofoundation/ixo-blockchain/x/payments/types"
```
Comment (migueldingli1997) [inline x/payments/keeper/subscriptions.go:6]:
```suggestion
	"github.com/ixofoundation/ixo-blockchain/x/payments/types"
```

Comment (migueldingli1997) [reply x/payments/keeper/subscriptions.go:6]: Change any `types2.` to `types.`

File: cmd/ixod/cmd/root.go (hunk 1 above)
Comment (stefaniatadama) [reply cmd/ixod/cmd/root.go:81]: This concerns the `ixod migrate` command. Running `ixod migrate` only migrates the cosmos-sdk modules to the requested version but not the ixo specific modules. We can leave it as is (and remove the comment) unless we want to incorporate our genesis migration script to also migrate ixo modules.

File: cmd/ixod/cmd/root.go (hunk 2 above)
Comment (stefaniatadama) [reply cmd/ixod/cmd/root.go:179]: I think we can remove the comment.

File: app/app.go (hunk 3 above)
Comment (stefaniatadama) [reply app/app.go:658]: Remove TODO.

File: x/did/types/codec.go (hunk 12 above)
Comment (stefaniatadama) [reply x/did/types/codec.go:21]: These were here before, do we remove them?

File: app/genesis.go [hunk 15]
```diff
@@ -14,6 +15,7 @@ import (
 type GenesisState map[string]json.RawMessage
 
 // NewDefaultGenesisState generates the default state for the application.
-func NewDefaultGenesisState() GenesisState {
-	return ModuleBasics.DefaultGenesis()
+func NewDefaultGenesisState(cdc codec.JSONMarshaler) GenesisState {
+	//encCfg := MakeTestEncodingConfig()
```
[the remark this answers is not in the document]
Comment (stefaniatadama) [reply app/genesis.go:19]: Gaia has the commented version, cosmos-sdk has the actual version.

[the remark this answers is not in the document]
Comment (stefaniatadama) [reply app/genesis.go:19]: We will remove the comment.

File: cmd/ixod/cmd/root.go (hunk 1 above)
Comment (stefaniatadama) [reply cmd/ixod/cmd/root.go]: Edited the comment.

**Pull request closed without merging**